subsys/net/ip/net_core.c

```diff
@@ -393,6 +396,9 @@ int net_recv_data(struct net_if *iface, struct net_pkt *pkt)
 	net_pkt_set_overwrite(pkt, true);
 	net_pkt_cursor_init(pkt);
 
+#if IS_ENABLED(CONFIG_NET_DSA)
+	iface = dsa_recv_set_iface(iface, &pkt);
```

Comment:
not sure it's the right place to do this. dsa is ethernet specific, and as such maybe ethernet L2 should be doing this

Reply:
The idea here is to read LAN port number (lan1,2,3) as soon as possible - i.e. from the raw buffer before any high level network code is executed.
For example here we read the LAN interface number (from appended tag by KSZ8794) before the struct iface is set for the packet. IMHO this is the best conceptual place to avoid issues latter.

Reply:
no architecturally this is by far the worst place: net_recv_data is a very generic function and you should not add L2 specific use cases there.

[edit] all DSA related stuff, besides the drivers,  must land in subsys/net/l2/ethernet

Reply:
The rationale to place `dsa_recv_set_iface()` function in net_core.c's `net_recv_data()` function was to set proper `struct iface` just before it is used in `net_pkt_set_iface()`. Afterwards the packet is passed to `net_queue_rx()`.

When moving the `dsa_recv_set_iface()` to `ethernet_recv()`, in the DSA case, we would have the iface set to master
port (which is NOT correct) when the packet enters `ethernet_recv()` (in net/l2/ethernet/ethernet.c). Would it be safe to have such mismatch?

Reply:
If you set the interface properly already in the driver before calling `net_recv_data()`, there is no need to have any changes in net_core.c

Reply:
I agree. I've moved this code to `eth_rx()` callback in eth_mcux.c

Reply:
After some excessive testing it turned out that the dsa_recv_set_iface() shall
be called in net_recv_data() (subsys/net/ip/net_core.c) as after some basic
checks (if e.g. iface != NULL) we are sure that the iface is ready to handle
the packet.

Moreover, in the same place the code to assign original iface (when
CONFIG_NET_ROUTING is enabled) is executed and hence the deduction that
DSA shall follow the same pattern).